io/io/src/TFile.cxx

```diff
@@ -823,7 +832,7 @@ void TFile::Init(Bool_t create)
       }
    }
 
-   {
+   if (!fLocallyManaged) {
```

Comment:
Removing:
```
   gROOT->GetUUIDs()->AddUUID(fUUID,this);
```
induces a non-ownership functional change.  Not registered the file UUID will make resolving `TRef` within the file non-functional.

Reply:
Ok, noted.  I think the use of the global lock can probably be avoided in TProcessUUID in favor of a local one.  Could we safely skip the Add and remove of the UUID for the case where ```fNProcessIDs == 0```?

Reply:
By definition no, I think. The add should increment fNProcessIDs.  

>  I think the use of the global lock can probably be avoided in TProcessUUID in favor of a local one. 

Great (by watch out for hidden/indirect risk of dead-lock: if the section with the local lock has any possibility of taking the global lock (indirectly, for example via a call to Print or a GetClass) then there is a risk of dead-lock (risk == guarantee that it will happen somewhere somehow).

Reply:
What I meant was that if fNProcessIDs for the file is 0 (before ```AddUUID```) doesn't this mean that there were no TRefs stored in this file?

Reply:
In fact I'm still not clear on the issue exactly.  Wouldn't skipping the UUID registration only affect TRefs which point directly to the TFile object itself?

Reply:
Sorry I was too quick to answer.  I think you are right.  Skipping the TFile's UUID registration would only affect the TRefs that are pointing to the TFile.  Furthermore, if `TFile::fNProcessIDs` is zero there should not be any `TRef` inside the file ... However `fNProcessIDs` is currently set after the UUID is registered ... but it should be fine to move it before the UUID registration.

Reply:
So actually I'm not sure the ```TFile::fNProcessIDs``` check is relevant in that case, since writing a TRef which points to an object with a UUID does not trigger the writing of the TProcessID.

Also the kIsReferenced bit for the TFile object doesn't seem to be preserved when a file is written and then opened again, so I'm not sure this case can be easily detected.

For the case of a TRef stored in a file and pointing to the same file we could probably handle the needed registration in the TRef streamer while reading.  This wouldn't handle the case of a TRef pointing to a file which is read from a different file or assigned during the  same process though...

Reply:
> This wouldn't handle the case of a TRef pointing to a file which is read from a different file or assigned during the same process though...

That's (one of) the real problem(s) ...  So I can't see how to avoid having to register the UUID ... and we probably need to find ways to make the adding more thread friendly.